kulkunen/models.py

```diff
@@ -0,0 +1,387 @@
+
+
+class AccessControlGrantQuerySet(models.QuerySet):
+    def active(self):
+        m = self.model
+        return self.filter(state__in=(m.REQUESTED, m.INSTALLING, m.INSTALLED, m.REMOVING))
+
+
+class AccessControlGrant(models.Model):
+    REQUESTED = 'requested'
+    INSTALLING = 'installing'
+    INSTALLED = 'installed'
+    CANCELLED = 'cancelled'
+    REMOVING = 'removing'
+    REMOVED = 'removed'
+    STATE_CHOICES = (
+        (REQUESTED, _('requested')),
+        (INSTALLED, _('installed')),
+        (CANCELLED, _('cancelled')),
+        (REMOVING, _('removing')),
+        (REMOVED, _('removed')),
+    )
+
+    user = models.ForeignKey(
+        AccessControlUser, related_name='grants', null=True, blank=True, on_delete=models.SET_NULL
+    )
+    resource = models.ForeignKey('AccessControlResource', related_name='grants', on_delete=models.CASCADE)
+
+    # If a Respa reservation is deleted, it will be marked as None here.
+    # AccessControlReservation with reservation == None should be deleted.
+    reservation = models.ForeignKey(
+        'resources.Reservation', on_delete=models.SET_NULL, null=True, related_name='access_control_grants'
+    )
+    starts_at = models.DateTimeField(null=True, blank=True)
+    ends_at = models.DateTimeField(null=True, blank=True)
+
+    # These are set if the grants are to be installed and removed at a later time.
+    install_at = models.DateTimeField(null=True, blank=True)
+    remove_at = models.DateTimeField(null=True, blank=True)
+
+    created_at = models.DateTimeField(auto_now_add=True)
+    removed_at = models.DateTimeField(auto_now_add=True)
+
+    access_code = models.CharField(verbose_name=_('access code'), max_length=32, null=True, blank=True)
+    state = models.CharField(
+        max_length=20, choices=STATE_CHOICES, default=REQUESTED, editable=False,
+        help_text=_('State of the grant')
+    )
+    identifier = models.CharField(
+        max_length=100, null=True, blank=True, verbose_name=_('identifier'),
+        help_text=_('Identifier of grant in the access control system (if any)')
+    )
+    installation_failures = models.PositiveIntegerField(
+        default=0, editable=False, help_text=_('How many times the system has tried to install this grant and failed')
+    )
+    removal_failures = models.PositiveIntegerField(
+        default=0, editable=False, help_text=_('How many times the system has tried to remove this grant (and failed)')
+    )
+
+    driver_data = JSONField(null=True, blank=True)
+
+    objects = AccessControlGrantQuerySet.as_manager()
+
+    class Meta:
+        index_together = (('resource', 'state'),)
+
+    def __str__(self) -> str:
+        # time_range = format_dt_range(settings.LANGUAGES[0][0], self.starts_at, self.ends_at)
+        return _("{user} {reservation} ({state})").format(
+            user=self.user, reservation=self.reservation, state=self.state
+        )
+
+    def is_active(self):
```

Comment:
`AccessControlGrantQuerySet.active()` has different rules compared to `AccessControlGrant.is_active()`. If they intentionally do different things, then the naming is confusing and could be changed to differentiate the results.